c++/src/kj/async-inl.h

```diff
@@ -2063,8 +2063,39 @@ concept NoWaitScope = !isSameType<Decay<T>, WaitScope>();
 // Define a Concept to use in our `coroutine_traits` specialization to validate allowable coroutine
 // parameter types.
 
+struct DisallowedWithinCoroutine {};
+
+template <typename T, typename = int>
+struct isDisallowedInCoroutine {
+  static constexpr bool value = false;
+};
+
+template <typename T>
+struct isDisallowedInCoroutine<T,
+    decltype((void) ((instance<T>()))->__kj_disallowed_in_coroutine, 0)> {
+  static constexpr bool value = true;
+};
+
+template <typename T>
+struct isDisallowedInCoroutine<T, decltype((void) Decay<T>::__kj_disallowed_in_coroutine, 0)> {
+  static constexpr bool value = true;
+};
+
 }  // namespace kj::_
 
+#define KJ_DISALLOW_IN_COROUTINE() \
```

Comment:
Bikeshedding:
* Maybe `KJ_DISALLOW_AS_COROUTINE_PARAM`? This more clearly communicates to the reader what is being disallowed -- it's still entirely possible to use the type _inside_ a coroutine, of course (and we can't prevent that unfortunately).
* I'd remove the `()` -- I don't think declare any other macros with empty args in KJ.